lib/index.d.ts

```diff
@@ -143,37 +147,39 @@ export function isBoom(obj: unknown, statusCode?: number): obj is Boom;
 
 
 /**
-* Specifies if an error object is a valid boom object
+* Applies options to an existing boom object, or creates a new boom object with the error as `cause`
 *
-* @param err - The error object to decorate
+* @param err - The target object
 * @param options - Options object
 *
-* @returns A decorated boom object
+* @returns A boom object
 */
-export function boomify<Data, Decoration>(err: Error, options?: Options<Data> & Decorate<Decoration>): Boom<Data> & Decoration;
-
+export function boomify<Data, Decoration>(err: Boom<Data>, options?: BoomifyOptions<Data> & { decorate: Decoration }): Boom<Data> & Decoration;
+export function boomify<Data, Decoration>(err: unknown, options?: BoomifyOptions<Data> & { decorate: Decoration }): Boom<Data> & Decoration;
+export function boomify<Data extends object>(err: Boom<Data>, options?: BoomifyOptions<Data>): Boom<Data>;
+export function boomify<Data>(err: unknown, options?: BoomifyOptions<Data>): Boom<Data>;
```

Comment:
The two last options might be superfluous here

```suggestion
export function boomify<Data, Decoration>(err: Boom<Data>, options?: BoomifyOptions<Data> & { decorate?: Decoration }): Boom<Data> & Decoration;
export function boomify<Data, Decoration>(err: unknown, options?: BoomifyOptions<Data> & { decorate?: Decoration }): Boom<Data> & Decoration;
```

[playground link](https://www.typescriptlang.org/play/?#code/C4TwDgpgBA8grsMCoF4pwHYGsMHsDuGAUERAB5i4BOwUAlhsBFQGYCGAxtDGMHbhgDOAHgAibYGwB8UAN5EoiqAHoAVKoVKoqqABUAFtAASu3QAUogycDiCoHXABMImpRq3aoAAWfs4AG1oAVgAGENdFVWUIqgg2RwF-EEtrWwBhJwgAfgAuKAw4AFsAI2YAbhItNXctHQBBR0c6PgE2fyhmKmp6DBZqQol+Yg8omLiEjCSoRwk2XKhxSQqI6ojPOowoXGKAKwgOWnxm-Q6yYCo2KDAuyBo6CDtgXEsIWgEoYEMOqi6qLd39sA1qMtLF4olks4HBcmPNZFAANpYCAgPJWKgMADmAF08mwMMkAL7LKrqYFQDb-PYHewCSQMLFQfHJEzmKCGeLMOz4QyxDqcE7I5J0OyXDnOP4YNiFaD4xxQABubTg0BFHy+4uYtMYEEYwOioPGEPZ4y5cKgYImUwRmqoaPOWNxKQxGExUAAPhajZNkuisQjsR78kVSn9iSQof42HyOFHBHYAEK4XCFMSzVBMgkychMDCOOwAUR+3XkpJqbigaTBTFF+Qg+CgSZTVMB6EEjM+0GuuAVdGc8pl8bYmOg3SLv31EQcQnOcAO1AAFIPBMPsvaXZiADRbXhDQTzHgtIRpyRSACUJKUqxGUAAqnnmElGRw2LZoH0-p3G8nCt8JyMDSUF833mTAcAIDBL0Ua9akrWwnl-TpuhmSQoCOT4mUaZohjaHoPwGI9LEgDg6BYOgOA+Z4vyQz9wBcACIhQtg8kWNgoJUMkbxFJtfwAWg+KgVW3BgmhfGt1TVNVLh4lsaQYKx8S4AA6SdDXBH16EEHi8mKZN-DiSDKivTjYIyDAFV1e4MC4KBdNwfT8R6UTBldFIJFsWlnCgKQ0FCcIGLUq1hUEABlZgLLtWy9IM9iYIrAxoBoqBl1XIgvC8KBeIAZiyzcAEYAHYoAAaiygAOfKACZvAy8hKBoew40TH8T0uNBmWzM5dXzKBxxLVSlFwBAkGAPJ4EQBBYpMisQuI0j7jsUjMz-bptmpWgpMVNo+yi5s1sBYFyS8MBo2laYIGKOA3X4updoczZPgkbceV1AShOmXAHnyXBaH0PtoC-ag6ExBhcL8la-hSkclIWCA-ECR5nnYfxBAgFSAqUWJ8IkBdnEuzF5js+6zzyBVcD7CpCRIOrqA2nVWE4aAeNIkBDz3VrOtzHq2YEERWJkUtjPLSJeuLSGHhXEdnSxQ6bx8OHX0Cb6MHo2pAMUS1jShtdpddKbhc8ABJFgoGR1Ht2on4ri6Xt+00plvz2gEDm3OUHYUmxBAyLzum1pk+W7W2IEcC2viVfwVVFPlgbwWJHFl2D5fh2hZ1Vtx1a9dSph7Tp-sJ6L8UpkgiEjaN30wA4hl2woWda7dRH2ahBgEKQF06PIeI57dcF3Xn5mZlhWd749+agAAyORzuhCQdYb6fCMJEnHdTUeJ7npujwqUu+RYCvCLsmvB7r2H56GVv2-QbA8EIbvh-3DufxZnmR9mGQJ-hKEN9nxuYSrxeH5TBzceJ8N5DAqMoZQ29y7WX3o-I+rFThczsPtA458fgAJXq-W+R577LyfnfDmS9O783AZA-YUYd57yrgfWu-M27oMvuBG+O4cH9zgUPHBhCMEc2WNOKwptkwZhoYPBcKsGx9SoAuM8Z4SB8NoH0XA1U0DCJAKI+sotfhSO3B-WYeR4SGAgHkAARPoNgbAjFQEJIoReRA5ECNwFlIR7C1HiLFloyen8YSGMngovIAByYo0Y-GWJsXYhRAAWJxKYWYuI0YuM82jpi6MngY4xpjzGWO3J4meej7H+MCVQYJhJF5AA)

Reply:
Also note that this notation does not properly capture what happens when `err` is `Boom<SomeData>` and options contains `{ data: SomeOtherData }`:

```ts
const foo = boomify(new Error() as Boom<{ foo: 'bar' }>, { data: { foo: "baz" } })
// Boom<{ foo: string; }>
```

Reply:
Also note that if `Decoration` contains a `data` property, it will get overtidden. We probably need to return something like `Boom<Data> & Omit<Decoration, 'data' | others?>`

Reply:
Thanks, I made a revised version that also preserves any subclassed Boom and doesn't override the `data` option.

I still include a Decoration-less option to avoid `Decoration` being inferred to `Boom.Boom<any>`, as well as explicit typing of the new `Terr` generic.

Reply:
awesome